ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll

```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
-  dependsOnSummaryComponentStackCons(callable, _, stack)
-}
+module SummaryTypeTrackerInput implements SummaryTypeTracker::Input {
+  // Dataflow nodes
+  class Node = DataFlow::Node;
 
-pragma[nomagic]
-private predicate dependsOnSummaryComponentStackCons(
-  SummarizedCallable callable, SummaryComponent head, SummaryComponentStack tail
-) {
-  dependsOnSummaryComponentStack(callable, SCS::push(head, tail))
-}
+  // Content
+  class TypeTrackerContent = DataFlowPublic::ContentSet;
 
-pragma[nomagic]
-private predicate dependsOnSummaryComponentStackConsLocal(
-  SummarizedCallable callable, SummaryComponent head, SummaryComponentStack tail
-) {
-  dependsOnSummaryComponentStackCons(callable, head, tail) and
-  not isNonLocal(head)
-}
+  class TypeTrackerContentFilter = ContentFilter;
 
-pragma[nomagic]
-private predicate dependsOnSummaryComponentStackLeaf(
-  SummarizedCallable callable, SummaryComponent leaf
-) {
-  dependsOnSummaryComponentStack(callable, SCS::singleton(leaf))
-}
+  TypeTrackerContentFilter getFilterFromWithoutContentStep(TypeTrackerContent content) {
+    (
+      content.isAnyElement()
+      or
+      content.isElementLowerBoundOrUnknown(_)
+      or
+      content.isElementOfTypeOrUnknown(_)
+      or
+      content.isSingleton(any(DataFlow::Content::UnknownElementContent c))
+    ) and
+    result = MkElementFilter()
+  }
 
-/**
- * Gets a data flow node corresponding to the local input or output of `call`
- * identified by `stack`, if possible.
- */
-pragma[nomagic]
-private DataFlow::Node evaluateSummaryComponentStackLocal(
-  SummarizedCallable callable, DataFlow::CallNode call, SummaryComponentStack stack
-) {
-  exists(SummaryComponent component |
-    dependsOnSummaryComponentStackLeaf(callable, component) and
-    stack = SCS::singleton(component) and
-    call.asExpr().getExpr() = callable.getACallSimple() and
-    result = evaluateSummaryComponentLocal(call, component)
-  )
-  or
-  exists(DataFlow::Node prev, SummaryComponent head, SummaryComponentStack tail |
-    prev = evaluateSummaryComponentStackLocal(callable, call, tail) and
-    dependsOnSummaryComponentStackConsLocal(callable, pragma[only_bind_into](head),
-      pragma[only_bind_out](tail)) and
-    stack = SCS::push(pragma[only_bind_out](head), pragma[only_bind_out](tail))
-  |
+  TypeTrackerContentFilter getFilterFromWithContentStep(TypeTrackerContent content) {
+    (
+      content.isAnyElement()
+      or
+      content.isElementLowerBound(_)
+      or
+      content.isElementLowerBoundOrUnknown(_)
+      or
+      content.isElementOfType(_)
+      or
+      content.isElementOfTypeOrUnknown(_)
+      or
+      content.isSingleton(any(DataFlow::Content::ElementContent c))
+    ) and
+    result = MkElementFilter()
+  }
+
+  // Summaries and their stacks
+  class SummaryComponent = FlowSummary::SummaryComponent;
+
+  class SummaryComponentStack = FlowSummary::SummaryComponentStack;
+
+  SummaryComponentStack singleton(SummaryComponent component) {
```

Comment:
Couldn't you just have used a predicate alias here?
`predicate SummaryComponentStack = FlowSummary::SummaryComponentStack::singleton/1`.

Reply:
I remember having some issues with using an alias, but I may have just been fumbling. I will try again :-)

Reply:
That worked fine...and looks a lot neater..